### PR TITLE
Corrected Note Type Field Mapping (#18131)

### LINKED ISSUE #18131: In "add note" changing note type should respect field name when possible
This bug is similar to https://github.com/ankidroid/Anki-Android/issues/18132 but considers a different view. However, it'll need a similar utility function. The people assigned to both tasks, if it's not the same person, should coordinator so that you reuse the code you can.

Ideally, you may want to first just create the utility function in a first PR, so that it can be used by the people working on both bug later

How to reproduce:
* create two note type named ABC and ACD, with fields A, B, and C, and A, C, D respectively
* open the add note editor
* choose note type ABC
* fill the fields with "A", "B" and "C"
* change the note type to ACD

Expected result:
* in anki desktop, the field contents is not "A", "C" and "B"

That is, when a field with name F exists in the original note type and the destination note type, then the content of the field A remains the same even if the position of F move. However, the content of the original field is not erased. Instead it's moved to another field. 

Actual result:
* The fields are still A, B, C

Please implement exactly the same algorithm to reorder the field content as upstream. You can find this algorithm in https://github.com/ankitects/anki/blob/83d0f5dae910ac03e950f7359bfac16842f3b59b/qt/aqt/addcards.py#L167

To be even more specific, what I want is:
* a function that, given the array of field's names of the previous and the next note type, return a data structure allowing to know which position move where, and unit tests ensuring that this really work as expected on a lot of simple cases and corner case
* a function that takes the note and the new note type and generate the note after this change of fields, and this should also come with unit tests (can be testing the same case as above)
* using this function in the add note view (note that the note type editor has a different interface, it shouldn't change)

I put it as "good first issue" because, while it may not be trivial from an algorithm point of view, it's a self contained, well defined problem.

### PR DESCRIPTION
**Pull request by Tushar4059x**
## Purpose / Description

This PR improves the field mapping logic when changing note types in the Note Editor. Previously, fields were mapped purely by position, which often led to data being placed in inappropriate fields when switching between note types with similar fields in different positions.

With this change, fields are now mapped by name first, then by position as a fallback. This means when users change a note type, fields with matching names will retain their content, regardless of their position in the note type.

### 1. Added `mapFieldsByNames` Method to NoteEditor class to map fields by name first:

```kotlin
    /**
     * Maps fields from old note type to new note type based on field names.
     *
     * @param oldNotetype The source note type
     * @param newNotetype The target note type
     * @return A mapping from old field index to new field index, matching fields by name when possible
     */
    private fun mapFieldsByNames(
        oldNotetype: NotetypeJson,
        newNotetype: NotetypeJson,
    ): Map<Int, Int> {
        // Get field maps for both old and new note types
        val oldFieldMap = Notetypes.fieldMap(oldNotetype)
        val newFieldMap = Notetypes.fieldMap(newNotetype)

        // Create a result map that will hold the mapping from old field indices to new field indices
        val resultMap: MutableMap<Int, Int> = HashUtil.hashMapInit(oldFieldMap.size)

        // Create a set to track which new fields have already been mapped to
        val mappedNewFieldIndices = mutableSetOf<Int>()

        // First, map fields with the same name
        for ((oldFieldName, oldFieldData) in oldFieldMap) {
            val oldFieldIndex = oldFieldData.first

            // Find the corresponding field in the new note type (if any)
            val newFieldData = newFieldMap[oldFieldName]
            if (newFieldData != null) {
                val newFieldIndex = newFieldData.first
                resultMap[oldFieldIndex] = newFieldIndex
                mappedNewFieldIndices.add(newFieldIndex)
            }
        }

        // For any old fields not mapped yet, try to map them to unused new fields with the same index (if possible)
        for (oldIndex in 0 until oldFieldMap.size) {
            if (!resultMap.containsKey(oldIndex)) {
                // Try to use the same index if it's available
                if (oldIndex < newFieldMap.size && !mappedNewFieldIndices.contains(oldIndex)) {
                    resultMap[oldIndex] = oldIndex
                    mappedNewFieldIndices.add(oldIndex)
                } else {
                    // Find any available index in the new note type
                    for (newIndex in 0 until newFieldMap.size) {
                        if (!mappedNewFieldIndices.contains(newIndex)) {
                            resultMap[oldIndex] = newIndex
                            mappedNewFieldIndices.add(newIndex)
                            break
                        }
                    }

                    // If we couldn't find any available field, map to the first field (some content is better than none)
                    if (!resultMap.containsKey(oldIndex) && newFieldMap.isNotEmpty()) {
                        resultMap[oldIndex] = 0
                    }
                }
            }
        }

        return resultMap
    }

```

### 2. Modified `EditNoteTypeListener.onItemSelected` Method  to use the new mapping logic:

**Before:**
```kotlin
val itemsLength = editorNote!!.items().size
modelChangeFieldMap = HashUtil.hashMapInit(itemsLength)
for (i in 0 until itemsLength) {
    modelChangeFieldMap!![i] = i
}
```

**After:**
```kotlin
val oldModel = currentEditedCard!!.noteType(getColUnsafe)
modelChangeFieldMap = mapFieldsByNames(oldModel, newModel).toMutableMap()
```

### 3. Added Comprehensive Test Cases in `NoteEditorFieldMappingTest`:

Created a new test class `NoteEditorFieldMappingTest` with multiple test scenarios to verify field mapping behavior.

**Example Test Case:**
```kotlin
@Test
fun testReorderedMatchingFields() {
    val oldNotetype = createNotetypeWithFields(listOf("Front", "Back", "Extra"))
    val newNotetype = createNotetypeWithFields(listOf("Back", "Extra", "Front"))

    val mapping = mapFieldsByNames(oldNotetype, newNotetype)

    assertEquals(2, mapping[0]) // Front -> index 2
    assertEquals(0, mapping[1]) // Back -> index 0
    assertEquals(1, mapping[2]) // Extra -> index 1
}
```

## Approach

1. Fields are first mapped by name to ensure logical consistency.
2. If no name match exists, the same index is used when possible.
3. Otherwise, the first available unmapped index is assigned.
4. As a last resort, unmatched fields are mapped to the first field.

## Example Use Case

For example, when switching from a note type with fields `["A", "B", "C"]` to `["A", "C", "D"]`:

- **Field A** content stays in **Field A** (matched by name).
- **Field C** content stays in **Field C** (matched by name).
- **Field B** content moves to **Field D** (only remaining unmapped field).

This prevents incorrect positional mapping where fields might be misplaced.

## Testing

Added unit tests in NoteEditorFieldMappingTest that cover all the scenarios described above.

## Screeenshots
 Before & After Mapping Fix:
Screenshot 2025-03-25 at 5 49 14 PM
Screenshot 2025-03-25 at 5 48 59 PM
Closes #18131

### CODE REVIEW

**Comment by Tushar4059x** [comment]
Apologies for the mix-up! I accidentally committed a fix for another issue in this branch.

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 1]
```diff
@@ -2713,12 +2775,10 @@ class NoteEditor :
...
                     }
-                // Initialize mapping between fields of old model -> new model
-                val itemsLength = editorNote!!.items().size
-                modelChangeFieldMap = HashUtil.hashMapInit(itemsLength)
-                for (i in 0 until itemsLength) {
-                    modelChangeFieldMap!![i] = i
-                }
+                // Initialize mapping between fields of old model -> new model based on field names
+                val oldModel = currentEditedCard!!.noteType(getColUnsafe)
+                modelChangeFieldMap = mapFieldsByNames(oldModel, newModel).toMutableMap()
```
Comment (criticalAY) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:2780]: Return early in case old model and new model are same

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 2]
```diff
@@ -2665,6 +2665,68 @@ class NoteEditor :
...
+
+        // Create a set to track which new fields have already been mapped to
+        val mappedNewFieldIndices = mutableSetOf<Int>()
+
+        // First, map fields with the same name
+        for ((oldFieldName, oldFieldData) in oldFieldMap) {
+            val oldFieldIndex = oldFieldData.first
+
+            // Find the corresponding field in the new note type (if any)
+            val newFieldData = newFieldMap[oldFieldName]
```
Comment (criticalAY) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:2694]: You can use Kotlin’s safe call with let to eliminate the explicit null check

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 3]
```diff
@@ -2665,6 +2665,68 @@ class NoteEditor :
...
+                        if (!mappedNewFieldIndices.contains(newIndex)) {
+                            resultMap[oldIndex] = newIndex
+                            mappedNewFieldIndices.add(newIndex)
+                            break
+                        }
+                    }
+
+                    // If we couldn't find any available field, map to the first field (some content is better than none)
+                    if (!resultMap.containsKey(oldIndex) && newFieldMap.isNotEmpty()) {
+                        resultMap[oldIndex] = 0
```
Comment (criticalAY) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:2721]: We can reduce nesting here and replace the inner loop with `firstOrNull`?

**Review by criticalAY — CHANGES_REQUESTED** [review]
You can reduce the number of comments in your PR's code apart from the changes I requested

**Comment by Tushar4059x** [comment]
@criticalAY just updated the code with all the changes you mentioned ... please re-check.
Thanks for the feedback :)

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 4]
```diff
@@ -2665,6 +2665,53 @@ class NoteEditor :
...
+            newFieldMap[oldFieldName]?.let { newFieldData ->
+                val newFieldIndex = newFieldData.first
+                resultMap[oldFieldIndex] = newFieldIndex
+                mappedNewFieldIndices.add(newFieldIndex)
+            }
+        }
+
+        // For any old fields not mapped yet, trying to map them to unused new fields with the same index (if possible)
+        for (oldIndex in 0 until oldFieldMap.size) {
+            if (!resultMap.containsKey(oldIndex)) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:2699]: invert the if for clarity

File: AnkiDroid/src/test/java/com/ichi2/anki/NoteEditorFieldMappingTest.kt [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+        val noteTypeJson =
+            JSONObject().apply {
+                put("name", "Test Note Type")
+                put("id", 1L)
+                put("flds", fieldsArray)
+                put("tmpls", JSONArray())
+                put("type", 0)
+                put("sortf", 0)
+            }
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/NoteEditorFieldMappingTest.kt:55]: nit/implementer's choice: feel free to use a JSON string instead, probably more readable

File: AnkiDroid/src/test/java/com/ichi2/anki/NoteEditorFieldMappingTest.kt [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+        return NotetypeJson(noteTypeJson)
+    }
+
+    /**
+     * Test implementation of mapFieldsByNames for testing
+     *
+     * This matches the implementation in NoteEditor class but is isolated for testing
+     */
+    private fun mapFieldsByNames(
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/NoteEditorFieldMappingTest.kt:65]: Consider: use `@VisibleForTesting` on the main method and make it public

File: AnkiDroid/src/test/java/com/ichi2/anki/NoteEditorFieldMappingTest.kt [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+package com.ichi2.anki
+
+import com.ichi2.libanki.NotetypeJson
+import org.json.JSONArray
+import org.json.JSONObject
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+class NoteEditorFieldMappingTest {
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/NoteEditorFieldMappingTest.kt:25]: Look into `JvmTest` so you don't need to isolate the code

**Review by david-allison — CHANGES_REQUESTED** [review]
Cheers! Code looks good, just a few improvements to the tests
